spec/models/spree/gateway/amazon_spec.rb

```diff
@@ -38,16 +51,148 @@
+      ).to_return(
+        return_values,
       )
+    end
 
-      auth = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
-      expect(auth).to be_success
+    context 'when approved' do
+      it 'succeeds' do
+        stub_auth_request(return_values: {
+          status: 200,
+          headers: {'content-type' => 'text/xml'},
+          body: build_mws_auth_approved_response(order: order),
+        })
+
+        response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+
+        expect(response).to be_success
+      end
+    end
+
+    context 'when declined' do
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/xml'},
+          status: 200,
+          body: build_mws_auth_declined_response(order: order),
+        })
+
+        response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+
+        expect(response).not_to be_success
+        expect(response.message).to eq('Authorization failure: InvalidPaymentMethod')
+      end
+    end
+
+    context 'with an ErrorResponse error' do
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/xml'},
+          status: 400,
+          body: build_mws_auth_error_response(order: order),
+        })
+
+        response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+
+        expect(response).not_to be_success
+        expect(response.message).to match(/^400 TransactionAmountExceeded:/)
+      end
+    end
+
+    context 'with a 5xx error' do
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/plain'},
+          status: 502,
+          body: 'Bad Gateway',
+        })
+
+        response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+
+        expect(response).not_to be_success
+        expect(response.message).to match(/502 Bad Gateway/)
+      end
+    end
+
+    # 500 is special-cased in the Amazon library
+    context 'with a 500 error' do
+      it 'fails' do
+        stub_auth_request(return_values: {
+          headers: {'content-type' => 'text/plain'},
+          status: 500,
+          body: 'Server Error',
+        })
+
+        expect {
+          response = payment_method.authorize(order.total, payment_source, {order_id: payment.send(:gateway_order_id)})
+        }.to raise_error(Spree::Core::GatewayError, 'InternalServerError')
+      end
+    end
+
+    # 503 is special-cased in the Amazon library
+    context 'with a 503 error' do
+      before do
+        # Without this the specs have a big pause while the Amazon gem retries
+        expect_any_instance_of(PayWithAmazon::Request).to receive(:get_seconds_for_try_count).at_least(:once).and_return(0)
```

Comment:
I'm wondering if this should be `allow_any_instance_of` instead. An expectation will break if `pay_with_amazon` changes their implementation, using `allow_any_instance_of` won't break but it might make the tests slower...